packages/datadog-instrumentations/src/knex.js

```diff
@@ -18,3 +22,58 @@ function patch (file) {
     return Builder
   })
 }
+
+addHook({
+  name: 'knex',
+  versions: ['>=2'],
+  file: 'lib/knex-builder/Knex.js'
+}, Knex => {
+  shimmer.wrap(Knex.Client.prototype, 'raw', raw => function () {
+    if (!startRawQueryCh.hasSubscribers) {
+      return raw.apply(this, arguments)
+    }
+
+    const sql = arguments[0]
+
+    // Skip query done by Knex to get the value used for undefined
+    if (sql === 'DEFAULT') {
+      return raw.apply(this, arguments)
+    }
+
+    const asyncResource = new AsyncResource('bound-anonymous-fn')
+
+    function onFinish () {
+      asyncResource.bind(function () {
+        finishRawQueryCh.publish()
+      }, this).apply(this)
+    }
+
+    startRawQueryCh.publish({ sql })
+
+    const rawResult = raw.apply(this, arguments)
+    wrapThenRaw(rawResult.then, onFinish, asyncResource)
+
+    return rawResult
+  })
+  return Knex
+})
+
+function wrapThenRaw (origThen, onFinish, asyncResource) {
+  return function then (onFulfilled, onRejected) {
+    arguments[0] = wrapCallback(asyncResource, onFulfilled, onFinish)
+    arguments[1] = wrapCallback(asyncResource, onRejected, onFinish)
+
+    return origThen.apply(this, arguments)
+  }
+}
+
+function wrapCallback (asyncResource, callback, onFinish) {
+  if (typeof callback !== 'function') return callback
```

Comment:
I think that you should wrap or create a function for error and success, always.

If customer is subscribing to `then` only with success method, if an error happens, you are not going to be able to launch `onFinish` event.

Reply:
With the current approach, only the callbacks provided in `then` are instrumented.
Besides that, `catch` and `asCallback` have been instrumented to ensure finish message is sent.